File: frontend/check-conversation.mjs
import { chromium, expect } from '@playwright/test'
const browser=await chromium.launch({channel:'chrome',headless:true})
try{
  const context=await browser.newContext({viewport:{width:1440,height:1000}})
  const page=await context.newPage()
  const errors=[],requests=[]
  page.on('pageerror',error=>errors.push(error.message))
  let mode='normal'
  await context.route('**/api/chat/stream',route=>{
    const body=route.request().postDataJSON();requests.push(body)
    const answer=mode==='empty'?'':mode==='thinking'?'<think>PRIVATE THOUGHT</think>':`Activity ${requests.length}: use paper strips for fraction practice, keeping the original ten-minute limit and no-calculator constraint. Model one example, ask the student to explain a second example, and end with a short independent check.`
    const result={answer,sources:[],model:'Test local model',retrieval:{query:body.question,student_ids:['STU-1020'],skill_ids:['S06'],retrieval_ms:1,matches:[]},metrics:{cached:false,total_ms:10,first_token_ms:5}}
    const frames=[['status',{message:'Preparing a complete answer…'}],['delta',{text:answer}],['done',result]]
    return route.fulfill({contentType:'text/event-stream',body:frames.map(([event,data])=>`event: ${event}\ndata: ${JSON.stringify(data)}\n\n`).join('')})
  })
  await page.goto('http://127.0.0.1:5178')
  const first='Help Julian with fractions. Ten minutes, no calculators. '+('Please use paper strips and keep practice concrete. '.repeat(15))
  for(const question of [first,'Make an activity.','Give two problems.','Adapt the second problem.','What did I originally ask for?','Revise the activity again.']){
    await page.getByRole('textbox',{name:'Ask about your students'}).fill(question)
    await page.getByRole('button',{name:'Send message'}).click()
    await expect(page.getByRole('button',{name:'New conversation'})).toBeEnabled()
  }
  expect(requests.at(-1).history).toHaveLength(10)
  expect(requests.at(-1).history[0].content).toBe(first.trim())
  expect(requests.at(-1).history[1].scope).toEqual({student_ids:['STU-1020'],skill_ids:['S06']})
  await expect(page.locator('.message.assistant .markdown')).toHaveCount(6)
  for(const failureMode of ['empty','thinking']){
    mode=failureMode
    await page.getByRole('textbox',{name:'Ask about your students'}).fill('Another useful activity.')
    await page.getByRole('button',{name:'Send message'}).click()
    await expect(page.getByRole('alert')).toBeVisible()
    await expect(page.locator('.markdown').filter({hasText:'PRIVATE THOUGHT'})).toHaveCount(0)
    expect(requests.at(-1).history.every(m=>m.content.trim())).toBe(true)
    const failedQuestionCount=await page.locator('.message.user').count()
    mode='normal'
    await page.getByRole('button',{name:'Retry',exact:true}).click()
    await expect(page.getByRole('button',{name:'New conversation'})).toBeEnabled()
    await expect(page.getByRole('alert')).toHaveCount(0)
    expect(await page.locator('.message.user').count()).toBe(failedQuestionCount)
    expect(requests.at(-1).history.every(m=>m.content!=='Another useful activity.'||m.role==='assistant'||failureMode==='thinking')).toBe(true)
  }
  await page.getByRole('button',{name:'New conversation'}).click()
  await expect(page.locator('.message')).toHaveCount(0)
  await page.getByRole('textbox',{name:'Ask about your students'}).fill('Start fresh with Maya.')
  await page.getByRole('button',{name:'Send message'}).click()
  await expect(page.getByRole('button',{name:'New conversation'})).toBeEnabled()
  expect(requests.at(-1).history).toEqual([])
  expect(errors).toEqual([])
  console.log('Full-history transport, retrieval scope, hidden invalid output, retry without duplicate turns, and conversation reset passed.')
}finally{await browser.close()}

File: scripts/check_conversation.py
"""Exercise a real local-model conversation; no generation or embedding mocks."""
import json
import os
import re
import time
from pathlib import Path
import httpx
from backend.conversation import answer_issue

base=os.getenv('FORMA_BASE_URL','http://127.0.0.1:8008').rstrip('/')
questions=[
    'How is Julian doing with fractions? I have ten minutes and no calculators. Suggest a focused teaching approach.',
    'Turn that into an activity with two practice problems and their answers.',
    'Make the second problem easier and explain how to use it.',
    'What constraints did I give you at the beginning, and how does the activity respect them?',
    'What about Noah with the same skill? How would you adapt the activity for him?',
]
history=[]
results=[]
with httpx.Client(timeout=240) as client:
    for turn,question in enumerate(questions,start=1):
        started=time.monotonic()
        result=None
        visible=''
        with client.stream('POST',base+'/api/chat/stream',json={'question':question,'history':history}) as response:
            response.raise_for_status()
            event=''
            for line in response.iter_lines():
                if line.startswith('event:'):event=line[6:].strip()
                if not line.startswith('data:'):continue
                data=json.loads(line[5:])
                if event=='status':print(json.dumps({'turn':turn,'status':data['message']}),flush=True)
                if event=='error':raise AssertionError(data)
                if event=='delta':visible+=data['text']
                if event=='done':result=data
        assert result and visible==result['answer']
        assert not result['metrics'].get('fallback'),result['answer']
        assert answer_issue(visible,question,bool(history),'stop') is None
        assert not re.search(r'<(?:think|analysis)>|the user (?:asks|wants)|we need to answer',visible,re.I)
        assert result['retrieval']['student_ids']==(['STU-1002'] if turn==5 else ['STU-1020'])
        assert result['retrieval']['skill_ids']==['S06']
        if turn==4:
            assert re.search(r'\b(?:ten|10)\b',visible,re.I) and 'calculator' in visible.lower(),visible
        record={'turn':turn,'question':question,'answer':visible,'seconds':round(time.monotonic()-started,1),
                'metrics':result['metrics'],'conversation':result['conversation'],'scope':{'student_ids':result['retrieval']['student_ids'],'skill_ids':result['retrieval']['skill_ids']}}
        results.append(record)
        print(json.dumps(record),flush=True)
        history.extend([{'role':'user','content':question},{'role':'assistant','content':visible,'scope':record['scope']}])
        Path('/private/tmp/forma-conversation-live.json').write_text(json.dumps(results,indent=2))
print('Five consecutive RAG conversation turns passed.',flush=True)
